primitives/npos-elections/src/balancing.rs

```diff
@@ -0,0 +1,202 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Balancing algorithm implementation.
+//!
+//! Given a committee `A` and an edge weight vector `w`, a balanced solution is one that
+//!
+//! 1.  it maximizes the sum of member supports, i.e `Argmax { sum(support(c)) }`. for all `c` in
+//! `A`.
+//! 2. it minimizes the sum of supports squared, i.e `Argmin { sum(support(c).pow(2)) }` for all `c`
+//! in `A`.
+//!
+//! See [`balance`] for more information.
+
+use crate::{IdentifierT, Voter, ExtendedBalance};
+use sp_arithmetic::traits::Zero;
+use sp_std::prelude::*;
+use sp_std::{cell::RefCell, rc::Rc};
+
+/// Balance the weight distribution of a given `voters` at most `iterations` times, or up until the
+/// point where the biggest difference created per iteration of all stakes is `tolerance`. If this
+/// is called with `tolerance = 0`, then exactly `iterations` rounds will be executed, except if no
+/// change has been made (`difference = 0`).
+///
+/// In almost all cases, a balanced solution will have a better score than an unbalanced solution,
+/// yet this is not 100% guaranteed because the first element of a [`ElectionScore`] does not
+/// directly related to balancing.
+///
+/// Note that some reference implementation adopt an approach in which voters are balanced randomly
+/// per round. To advocate determinism, we don't do this. In each round, all voters are exactly
+/// balanced once, in the same order.
+///
+/// Also, note that due to re-distribution of weights, the outcome of this function might contain
+/// edges with weight zero. The call site should filter such weight if desirable. Moreover, the
+/// outcome might need balance re-normalization, see `Voter::try_normalize`.
+///
+/// ### References
+///
+/// - [A new approach to the maximum flow problem](https://dl.acm.org/doi/10.1145/48014.61051).
+/// - [Validator election in nominated proof-of-stake](https://arxiv.org/abs/2004.12990)
+pub fn balance<AccountId: IdentifierT>(
+	voters: &mut Vec<Voter<AccountId>>,
+	iterations: usize,
+	tolerance: ExtendedBalance,
+) -> usize {
+	if iterations == 0 { return 0; }
+
+	let mut iter = 0 ;
+	loop {
+		let mut max_diff = 0;
+		for voter in voters.iter_mut() {
+			let diff = balance_voter(voter, tolerance);
+			if diff > max_diff { max_diff = diff; }
+		}
+
+		iter += 1;
+		if max_diff <= tolerance || iter >= iterations {
+			break iter;
+		}
+	}
+}
+
+/// Internal implementation of balancing for one voter.
+pub(crate) fn balance_voter<AccountId: IdentifierT>(
+	voter: &mut Voter<AccountId>,
+	tolerance: ExtendedBalance,
+) -> ExtendedBalance {
+	// create a shallow copy of the elected ones. The original one will not be used henceforth.
+	let mut elected_edges = voter.edges
+		.iter_mut()
+		.filter(|e| e.candidate.borrow().elected)
+		.map(|e| Rc::new(RefCell::new(e)))
+		.collect::<Vec<_>>();
```

Comment:
I'm not sure why you use a RefCell here, it seems you never need inner mutability at this stage,

why no just using a `Vec<&mut Edge>`
```suggestion
	let mut elected_edges = voter.edges
		.iter_mut()
		.filter(|e| e.candidate.borrow().elected)
		.collect::<Vec<&mut _>>();
```
for this to compile you need to remove all borrow and borrow_mut on edge in the code later

Reply:
Excellent, it works great.